### DISCUSSION TITLE
Nested mapping does not utilize the mappers defined for the types of the nested properties

### DISCUSSION THREAD

**Discussion by pchasco** [riok/mapperly#274] [Q&A] [answered]
I have two classes. Each class has an associated mapper class. Class A has a property of type `IEnumerable<B>`. When I look at the source code, instead of using the mapper defined for class B, the mapper instead implements its own private mapping method for class B.

Mapper generated for class "B"
```C#
#nullable enable
namespace Mappers
{
    public partial class IdWithConcurrencyTokenMapper
    {
        public partial Dto.IdWithConcurrencyTokenDto ToDto(Entities.IdWithConcurrencyToken entity)
        {
            var target = new Dto.IdWithConcurrencyTokenDto();
            target.Id = entity.Id;
            target.ConcurrencyToken = entity.ConcurrencyToken;
            return target;
        }
    }
}
```

Mapper generated for class "A"
```C#
namespace Mappers
{
    public partial class PushChangesResultMapper
    {
        public partial Dto.PushChangesResultDto ToDto(Sync.PushChangesResult entity)
        {
            var target = new Dto.PushChangesResultDto();
            target.ModifiedDates = System.Linq.Enumerable.Select(entity.ModifiedDates, x => MapToIdWithConcurrencyTokenDto(x));
            return target;
        }

// This should be using the mapper that was defined manually
        private Dto.IdWithConcurrencyTokenDto MapToIdWithConcurrencyTokenDto(Entities.IdWithConcurrencyToken source)
        {
            var target = new Dto.IdWithConcurrencyTokenDto();
            target.Id = source.Id;
            target.ConcurrencyToken = source.ConcurrencyToken;
            return target;
        }
    }
}
```

Ideally, class A would use the mapper defined for class B rather than create its own mapper. Aside from code bloat, this is an issue when class B has custom mapping logic. In my case, this essentially rules out using Mapperly as my class "B" has custom mapping logic.

A further challenge here is that the documentation recommends creating a private mapping method that is invoked by a public method when requiring Before/After map functionality. Any solution to the above mentioned problem would need to know how to call that public method when it is mapping nested properties/fields.

**Comment by latonz** [comment, marked as answer]
Thank you for opening this discussion. I think what you are trying will work if you use the same Mapper class for mappings of class A and B or if the mapper for class B extends the mapper of class A. If you use separate mapper classes for mappings they don't now about each other.

**Comment by pchasco** [comment]
Thanks for the suggestion! That seems like it should work. The only issue
is for people who want to reduce coupling between their mappers, but in
most cases it should be fine.

**Comment by zeidoo** [comment]
Do you think implementing shared configurations is possible?

Since this project is heavily inspired by mapstruct, I'm talking about this feature in particular:
https://mapstruct.org/documentation/stable/reference/html/#shared-configurations

I guess, for now, a solution is simply:
```
[Mapper]
public partial class ClassAMapper{
  public partial A MapA(ADto a);

  private B MapB(BDto b) => new ClassBMapper().MapB(b);
}

[Mapper]
public partial class ClassBMapper{
  // whatever custom mapping is needed
  public partial B MapB(BDto b);
}
```

**Reply by latonz** [reply]
Created https://github.com/riok/mapperly/issues/513 to track this.
